### ISSUE TITLE
The BEST model is directly turned on the TEST set

### ISSUE THREAD

**Issue by juntaoy** [ShannonAI/mrc-for-flat-nested-ner#7]
Hi,

I've looked into the code in  run_mrc_ner.py, in the train() method (line 258- line 282):

if tmp_dev_f1 > dev_best_f1 or tmp_dev_acc > dev_best_acc:
    dev_best_acc = tmp_dev_acc 
    dev_best_loss = tmp_dev_loss 
    dev_best_precision = tmp_dev_prec 
    dev_best_recall = tmp_dev_rec 
    dev_best_f1 = tmp_dev_f1 

     tmp_test_loss, tmp_test_acc, tmp_test_prec, tmp_test_rec, tmp_test_f1 = eval_checkpoint(model, test_dataloader, config, device, n_gpu, label_list, eval_sign="test")
     print("......"*10)
     print("TEST: loss, acc, precision, recall, f1")
     print(tmp_test_loss, tmp_test_acc, tmp_test_prec, tmp_test_rec, tmp_test_f1)

     if tmp_test_f1 > test_best_f1 or tmp_test_acc > test_best_acc:
         test_best_acc = tmp_test_acc 
         test_best_loss = tmp_test_loss 
         test_best_precision = tmp_test_prec 
         test_best_recall = tmp_test_rec 
         test_best_f1 = tmp_test_f1 

          # export model 
          if config.export_model:
              model_to_save = model.module if hasattr(model, "module") else model 
              output_model_file = os.path.join(config.output_dir, "bert_finetune_model.bin")
              torch.save(model_to_save.state_dict(), output_model_file)

It seems to me that the model is saved based on the best test set performance, although you do this only when the development set performance also improves. But still, this approach compromises the results with the test set. So normally people only use the dev set for selecting the best model and only tested on the test set once when training finishes. 

I wonder can you also let us know the results on the best model selected only on the development set. I am sure you will have them in the log since every time the development performance improves the test set is evaluated:) (based on your script)

Best,

Juntao

**Comment by edchengg** [comment]
If we have a higher dev F1 but lower test F1, the model will not be saved?

**Comment by juntaoy** [comment]
From the code, yes, so I am waiting for the author to get confirmations, also hope they can provide the results that do not use the test set for tuning. Otherwise feels to me it is not a fair comparison with previous work. As it happens all the time that you got a batter development result but the test result went down. To tuning on the test set can easily get you a few points, so the results is always much higher. Also in a real-world scenario, you won't have the gold annotation for the test set:)
